addons/survey/models/survey_user.py

```diff
@@ -218,22 +218,20 @@ def _compute_answer_is_correct(self):
                 answer.answer_is_correct = False
 
     @api.constrains('skipped', 'answer_type')
-    def _answered_or_skipped(self):
-        for uil in self:
-            if not uil.skipped != bool(uil.answer_type):
-                raise ValidationError(_('This question cannot be unanswered or skipped.'))
+    def _check_answered_or_skipped(self):
+        if any(line.skipped == bool(line.answer_type) for line in self):
+            raise ValidationError(_('A question is either skipped, either answered. Not both.'))
 
     @api.constrains('answer_type')
     def _check_answer_type(self):
-        for uil in self:
-            fields_type = {
-                'char_box': bool(uil.value_char_box),
-                'numerical_box': (bool(uil.value_numerical_box) or uil.value_numerical_box == 0),
-                'date': bool(uil.value_date),
-                'text_box': bool(uil.value_text_box),
-                'suggestion': bool(uil.suggested_answer_id)
-            }
-            if not fields_type.get(uil.answer_type, True):
+        for line in self:
+            if line.answer_type == 'suggestion':
+                field_name = 'suggested_answer_id'
+            elif line.answer_type:
+                field_name = 'value_%s' % line.answer_type
+            else:
+                field_name = False
+            if field_name and not line[field_name]:
```

Comment:
Curiosity question here:

What will happen if it tries to do `line[False]` ?

Reply:
I don't think it holds a 'False' key and should crash. Anyway the "if field_name" is there for that.